src/main/kotlin/pt/up/fe/ni/website/backend/email/TemplateEmailBuilder.kt

```diff
@@ -0,0 +1,95 @@
+package pt.up.fe.ni.website.backend.email
+
+import com.samskivert.mustache.Mustache
+import org.commonmark.ext.front.matter.YamlFrontMatterExtension
+import org.commonmark.ext.front.matter.YamlFrontMatterVisitor
+import org.commonmark.parser.Parser
+import org.commonmark.renderer.html.HtmlRenderer
+import org.commonmark.renderer.text.TextContentRenderer
+import org.springframework.boot.autoconfigure.mustache.MustacheResourceTemplateLoader
+import org.springframework.core.io.Resource
+import org.springframework.core.io.UrlResource
+import org.springframework.mail.javamail.MimeMessageHelper
+import org.springframework.util.ResourceUtils
+import pt.up.fe.ni.website.backend.config.email.EmailConfigProperties
+
+abstract class TemplateEmailBuilder<T>(
+    private val template: String
+) : BaseEmailBuilder() {
+    private companion object {
+        val commonmarkParser: Parser = Parser.builder().extensions(
+            listOf(
+                YamlFrontMatterExtension.create()
+            )
+        ).build()
+        val commonmarkHtmlRenderer: HtmlRenderer = HtmlRenderer.builder().build()
+        val commonmarkTextRenderer: TextContentRenderer = TextContentRenderer.builder().build()
+    }
+
+    private var data: T? = null
+
+    fun data(data: T) = apply {
+        this.data = data
+    }
+
+    override fun build(helper: MimeMessageHelper, emailConfigProperties: EmailConfigProperties) {
```

Comment:
this function is a bit verbose, I'd consider splitting it into more methods if possible